src/clib/pioc.c

```diff
@@ -500,27 +501,38 @@ int PIOc_set_hint(const int iosysid, char hint[], const char hintval[])
 int PIOc_finalize(const int iosysid)
 {
   iosystem_desc_t *ios, *nios;
+  int msg;
+  int mpierr;
 
   ios = pio_get_iosystem_from_id(iosysid);
   if(ios == NULL)
-    return PIO_EBADID; 
-  /* FIXME: The memory for ioranks is allocated in C only for intracomms
-   * Remove this check once mem allocs for ioranks completely moves to the
-   * C code
-   */ 
-  if(ios->intercomm == MPI_COMM_NULL){
-    if(ios->ioranks != NULL){
-      free(ios->ioranks);
-    }
+    return PIO_EBADID;
+  
+  /* If asynch IO is in use, send the PIO_MSG_EXIT message from the
+   * comp master to the IO processes. */
+  if (ios->async_interface && !ios->comp_rank)
```

Comment:
I find !ios->comp_rank a confusing way to point to comp rank 0.  I wonder if we should introduce a constant - say COMP_MASTER_RANK set to 0 so that we have ios->comp_rank == COMP_MASTER_RANK ?   Do the same for IO_MASTER_RANK?